server/core_storage.go

```diff
@@ -651,11 +652,12 @@ func storagePrepBatch(batch *pgx.Batch, authoritativeWrite bool, op *StorageOpWr
 			UPDATE storage SET value = $4, version = $5, read = $6, write = $7, update_time = now()
 			WHERE collection = $1 AND key = $2 AND user_id = $3 AND version = $8
 		` + writeCheck + `
+			AND NOT (storage.version = $5 AND storage.read = $6 AND storage.write = $7) -- micro optimization: don't update row unnecessarily
```

Comment:
I don't think this optimization is applicable to OCC writes.  Assume db has version A and 2 concurrent transactions write version B , using A as OCC version condition. 

 If these 2 transactions are NOT overlapping in time, then second update will fail due to normal OCC version check.

If concurrent updates are overlapping, then one of them will block, waiting for another one to commit. Once it commits, it will recheck WHERE on the new row  and it will fail due to OCC check.

Only case this optimization can work is if versions A and B are the same version , that is app writes the same row as it read from DB before, which is very unlikely to be a common case and doesn't worth optimizing for

Reply:
This would indeed only apply if an object is re-written without applying any change to the last read version.

Whilst this shouldn't happen, I wouldn't be surprised if it does due to bad client logic at updating some shared object.

Even if unlikely, in the case it does happen it could save the db some work. Do you think it is not worth it regardless?

Reply:
without OCC support, whole change is probably just a single condition in non-OCC SQL query without even Go code changes.  Supporting OCC is more complicated with next to no benefit.